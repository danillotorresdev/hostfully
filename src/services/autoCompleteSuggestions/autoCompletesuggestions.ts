import { nominatimBaseAPI } from "@/services/nominatimBaseAPI";

type Place = {
  place_id: number;
  licence: string;
  osm_type: string;
  osm_id: number;
  lat: string;
  lon: string;
  category: string;
  type: string;
  place_rank: 10;
  importance: number;
  addresstype: string;
  name: string;
  display_name: string;
  boundingbox: string;
};

type Places = Place[];

export const placesAutoCompleteSuggestions = (search: string) =>
  nominatimBaseAPI.get<Places>("/search.php", {
    params: {
      city: search,
      format: "jsonv2",
    },
  });
