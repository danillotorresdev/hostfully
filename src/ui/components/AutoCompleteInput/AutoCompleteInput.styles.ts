import { styled } from "@linaria/react";

export const SuggestionsWrapper = styled.div`
  position: relative;
  border-radius: 20px;
  background-color: white;
`;

export const SuggestionsList = styled.ul`
  position: absolute;
  top: 100%;
  left: 0;
  width: 100%;
  border-top: none;
  list-style: none;
  padding: 0;
  margin: 0;
  z-index: 10;
  box-shadow: 0 0 5px 0 rgba(0, 0, 0, 0.1);
  display: block;

  button {
    width: 100%;
    padding: 0.5rem;
    text-align: left;
    border: none;
    background-color: white;
    cursor: pointer;
    border-bottom: 1px solid #f5f5f5;
    z-index: 999;
  }
`;
