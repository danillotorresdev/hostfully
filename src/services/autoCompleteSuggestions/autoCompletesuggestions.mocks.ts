import { Places } from "@/services/autoCompleteSuggestions/autoCompletesuggestions";

export const placesMock: Places = [
  {
    addresstype: "city",
    boundingbox: "48.8155750,48.9021560,2.2241990,2.4699210",
    category: "tourism",
    display_name: "Paris, Île-de-France, France",
    importance: 0.935,
    lat: "48.8566969",
    licence:
      "Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright",
    lon: "2.3514616",
    name: "Paris",
    osm_id: 7444,
    osm_type: "relation",
    place_id: 23424846,
    place_rank: 10,
    type: "city",
  },
  {
    addresstype: "town",
    boundingbox: "48.8155750,48.9021560,2.2241990,2.4699210",
    category: "tourism",
    display_name: "Paris, Île-de-France, France",
    importance: 0.935,
    lat: "48.8566969",
    licence:
      "Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright",
    lon: "2.3514616",
    name: "Paris",
    osm_id: 7444,
    osm_type: "relation",
    place_id: 23424846,
    place_rank: 10,
    type: "town",
  },
  {
    addresstype: "village",
    boundingbox: "48.8155750,48.9021560,2.2241990,2.4699210",
    category: "tourism",
    display_name: "Paris, Île-de-France, France",
    importance: 0.935,
    lat: "48.8566969",
    licence:
      "Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright",
    lon: "2.3514616",
    name: "Paris",
    osm_id: 7444,
    osm_type: "relation",
    place_id: 23424846,
    place_rank: 10,
    type: "village",
  },
  {
    addresstype: "municipality",
    boundingbox: "48.8155750,48.9021560,2.2241990,2.4699210",
    category: "tourism",
    display_name: "Paris, Île-de-France, France",
    importance: 0.935,
    lat: "48.8566969",
    licence:
      "Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright",
    lon: "2.3514616",
    name: "Paris",
    osm_id: 7444,
    osm_type: "relation",
    place_id: 23424846,
    place_rank: 10,
    type: "municipality",
  },
];
