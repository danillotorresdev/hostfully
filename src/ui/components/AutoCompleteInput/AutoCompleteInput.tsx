import { Input, InputProps } from "@/ui/components/Input/Input";
import { useAutoComplete } from "@/utils/hooks/useAutoComplete/useAutoComplete";
import * as S from "./AutoCompleteInput.styles";
import { useState } from "react";

type AutoCompleteInputProps = Omit<InputProps, "onChange"> & {
  onPlaceChange: (place: string) => void;
  defaultValue?: string;
};

export const AutoCompleteInput = ({
  onPlaceChange,
  defaultValue,
  ...props
}: AutoCompleteInputProps) => {
  const { handleSelect, inputValue, suggestions, handleChange } =
    useAutoComplete();

  const [inputValueState, setInputValueState] = useState(
    inputValue === "" && defaultValue ? defaultValue : undefined,
  );

  return (
    <div>
      <Input
        {...props}
        defaultValue={defaultValue}
        value={inputValueState}
        type="text"
        onChange={(e) => {
          handleChange(e.target.value);
          setInputValueState(e.target.value);
          onPlaceChange(e.target.value);
        }}
      />
      <S.SuggestionsWrapper>
        <S.SuggestionsList>
          {suggestions.map((place) => (
            <li key={place}>
              <button
                onClick={() => {
                  handleSelect(place);
                  setInputValueState(place);
                }}
              >
                {place}
              </button>
            </li>
          ))}
        </S.SuggestionsList>
      </S.SuggestionsWrapper>
    </div>
  );
};
