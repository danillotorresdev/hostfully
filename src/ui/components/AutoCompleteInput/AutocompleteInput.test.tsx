import { AutoCompleteInput } from "./AutoCompleteInput";
import { beforeEach, describe, expect, it, vi } from "vitest";
import {
  act,
  fireEvent,
  render,
  screen,
} from "@/utils/settings/tests/utilities";
import * as placesAutoCompleteSuggestions from "@/services/autoCompleteSuggestions/autoCompletesuggestions";
import { placesMock } from "@/services/autoCompleteSuggestions/autoCompletesuggestions.mocks";
import { AxiosHeaders } from "axios";

describe("AutoCompleteInput component", () => {
  vi.useFakeTimers();
  beforeEach(() => {
    vi.clearAllMocks();
  });

  const placesAutoCompleteSuggestionsSpy = vi
    .spyOn(placesAutoCompleteSuggestions, "placesAutoCompleteSuggestions")
    .mockResolvedValue({
      data: placesMock,
      status: 0,
      statusText: "",
      headers: {},
      config: {
        headers: {} as AxiosHeaders,
      },
    });

  it("should match snapshot", () => {
    const { asFragment } = render(
      <AutoCompleteInput onPlaceChange={() => {}} label={"City"} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders input field and suggestions correctly", async () => {
    render(<AutoCompleteInput onPlaceChange={() => {}} label={"City"} />);

    const inputField = screen.getByRole("textbox");
    expect(inputField).toBeDefined();

    fireEvent.change(inputField, { target: { value: "New York" } });

    expect(screen.getByDisplayValue("New York")).toBeDefined();

    await act(async () => {
      vi.advanceTimersByTime(500);
    });

    expect(placesAutoCompleteSuggestionsSpy).toHaveBeenCalledWith("New York");
  });

  it("calls onPlaceChange callback when input value changes", () => {
    const onPlaceChange = vi.fn();
    render(<AutoCompleteInput onPlaceChange={onPlaceChange} label={"City"} />);

    const inputField = screen.getByRole("textbox");

    fireEvent.change(inputField, { target: { value: "New York" } });

    expect(onPlaceChange).toHaveBeenCalledWith("New York");
  });

  it("selects suggestion and updates input value when suggestion is clicked", async () => {
    render(<AutoCompleteInput onPlaceChange={() => {}} label={"City"} />);

    const inputField = screen.getByRole("textbox");

    fireEvent.change(inputField, { target: { value: "Paris" } });

    await act(async () => {
      vi.advanceTimersByTime(1000);
    });

    const suggestion = screen.getByRole("button", { name: "Paris, France" });

    fireEvent.click(suggestion);

    expect(screen.getByDisplayValue("Paris, France")).toBeDefined();
  });
});
