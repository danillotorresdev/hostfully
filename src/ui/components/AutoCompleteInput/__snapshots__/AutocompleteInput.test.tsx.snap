// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AutoCompleteInput component > should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="mocked-styled-0 i19zws4p"
    >
      <label
        class="mocked-styled-1 l5v2eyb"
        for="City"
      >
        City
      </label>
      <input
        class="mocked-styled-2 i1hv9n31"
        id="City"
        type="text"
        value=""
      />
    </div>
    <div
      class="mocked-styled-3 skesand"
    >
      <ul
        class="mocked-styled-4 s1qvsmxh"
      />
    </div>
  </div>
</DocumentFragment>
`;
